### ISSUE TITLE
Reading input from filename

### ISSUE THREAD

**Issue by Mike-DE-RE** [RWAP/PrinterToPDF#2]
Hi, me again. Hope, you don't regret to have released this project to public. :-)

As described in your source, "read_byte_from_printer" must be adopted to capture the input.

To see how your program works, I want to take input from a file.

Due to my lack of knowledge regarding "C", I took your openfont() function as a sample to read a hard coded file "./Test1.prn".

int read_byte_from_printer(char *filename)
{
    unsigned char databyte;
    char chr = 'A';
    int i = 0;
    FILE *inputfile;
    inputfile = fopen("./Test1.prn", "r");
    if (inputfile == NULL) return -1;
    while (!feof(inputfile)) {
        chr = fgetc(inputfile);
        databyte = databyte + chr;
        i++;
    }
    *byte = databyte;
    fclose(inputfile);
    return 1;
}

I get some errors compiling

gcc PrinterConvert.c -o example -lSDL -lrt

PrinterConvert.c: In function ‘read_byte_from_printer’:
PrinterConvert.c:111: error: ‘byte’ undeclared (first use in this function)
PrinterConvert.c:111: error: (Each undeclared identifier is reported only once
PrinterConvert.c:111: error: for each function it appears in.)

Is this the right way?

**Comment by RWAP** [comment]
Try using:

```
FILE *inputFile;
int initialize()
{
    /* routine could be used here to open the input file or port for reading 
    *  example is for reading from an input file called ./Test1.prn
    *  The routine is not error trapped at present
    */
    inputFile = fopen("./Test1.prn", "r");
    if (inputFile == NULL) return -1;
}

int read_byte_from_printer(unsigned char *bytex)
{
    /* This routine needs to be written according to your requirements
    * the routine needs to fetch the next byte from the captured printer data file (or port).
    */
    unsigned char databyte;
    // read databyte from file or port...
    // Example below is reading a character from the file opened in initialize();
    // take account of msbsetting : 0 - use bit 7 as set in data, 1 - clear bit 7 of all data, 2 - set bit 7 of all data

    while (!feof(inputFile)) {
        databyte = fgetc(inputFile);
        switch (msbsetting) {
        case 0:
            // No change
            break;
        case 1:
            // MSB is set byte 7 to 0
            databyte = databyte & 127;
            break;
        case 2:
            // MSB is set byte 7 to 1
            databyte = databyte | 128;
            break;
        }
        *bytex = databyte;
        return 1;
    }
    // No more data to be read from file
    fclose(inputFile);
    return 0;
}
```

**Comment by Mike-DE-RE** [comment]
Ok, had to delete initialize() from source as well, because it is contained in above snippet. Compiles without errors.

Running ./PrinterConvert

showes
Usage: ./PrinterConvert <timeout> <divisor> <font> <font_direction> <sdl> <path>

Usage: ./PrinterConvert 4 3 font2/SIEMENS.C16 1 sdlon /home/pi/data

timeout=4          --> if no char comes in within 4 sec, printout current page
divisor=30         --> reduce sdl display to 301763f original size
font=font2/SIEMENS.C16      --> rload this font in font memory area
font_direction=1   --> 0=lsb left 1=lsb right
sdl=sdlon          --> display printout in sdl window
path=/home/pi/data --> store all in this directory

Running

./PrinterConvert 4 3 ./SIEMENS.C16 1 sdlon ./

or

./PrinterConvert 4 3 ./ 1 sdlon ./

showes a "memory access error".

**Comment by Mike-DE-RE** [comment]
Maybe it's the wrong translation. Could as well be "segmentation fault".

**Comment by Mike-DE-RE** [comment]
Can the problem be caused by "sdlon"? Can i leave it away or do I have the option "sdloff"?

**Comment by RWAP** [comment]
To be honest, I am not sure.  If you do not include sdlon in the calling parameters, then the code will not copy the output to the screen.

I am waiting on Sascha (who co-wrote the original code with me) to have a look at this - it could be related to the large amount of memory we are setting aside for the 720 x 720 dpi image!

**Comment by Mike-DE-RE** [comment]
So allocating this amount of memory runs on the PI? How much RAM does the PI provide?

**Comment by RWAP** [comment]
The Raspberry Pi offers 1GB - it might use different memory allocation however, as it stores a lot on the SD card itself.  As I say, I need Sascha's input on this issue really

**Comment by RWAP** [comment]
Just a thought - try replacing line 32 with:

```
    unsigned char *printermemory;
    char = malloc (5955 * 8417 * 3); // Allow 3 bytes per pixel
    if (char == NULL) {
        fprintf(stderr, "Can't allocate memory for BMP file.\n");
        return (0);
    }

```
You would also need to alter lines 131 to 133:
```
    // No more data to be read from file
    fclose(inputFile);
    return 0;
```

should then read:

```
    // No more data to be read from file
    fclose(inputFile);
    free(printermemory);
    return 0;
```

The other option would be to treat each printed line separately - saving the current stored line to disk when the vertical print position alters, and then re-loading from disk if necessary.  This would slow down the processing of the page, but dramatically reduce the amount of memory required.  When the end of the page is reached, all of the files could be merged together (with a suitable header) to create the required bmp file for conversion to pdf.

**Comment by RWAP** [comment]
Try the latest version of the code - rather than using the malloc above, I have used a different approach to storing the printed information which has reduced the memory usage by  a factor of 3

**Comment by Mike-DE-RE** [comment]
I have tried remote. But the machine I want to run in on is off and I cannot switch it on from remote. So I have to try it tomorrow.

**Comment by Mike-DE-RE** [comment]
Ok, you learn something new every day. ;-)

So here comes today's lesson:

New version of the code compiles without errors. Program still not working.

After having switches the language to English I am sure about the correct translation: "Segmentation fault"

So I will have to learn how to debug a program under Linux to drive this further:

--- 1st try (start) ---

gdb PrinterConvert
GNU gdb 6.8
Copyright (C) 2008 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "i586-suse-linux"...
(gdb) run
Starting program: /home/user/PrinterToPDF/PrinterConvert
[Thread debugging using libthread_db enabled]
Usage: ./PrinterConvert <timeout> <divisor> <font> <font_direction> <sdl> <path>

Usage: ./PrinterConvert 4 3 font2/SIEMENS.C16 1 sdlon /home/pi/data

timeout=4          --> if no char comes in within 4 sec, printout current page
divisor=30         --> reduce sdl display to 301763f original size
font=font2/SIEMENS.C16      --> rload this font in font memory area
font_direction=1   --> 0=lsb left 1=lsb right
sdl=sdlon          --> display printout in sdl window
path=/home/pi/data --> store all in this directory

Program exited normally.
(gdb)

--- 1st try (end) ---

Looks fine so far.

--- 2nd try (start) ---

gdb PrinterConvert
GNU gdb 6.8
Copyright (C) 2008 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "i586-suse-linux"...
(gdb) run 4 3 SIEMENS.C16 1 sdlon ./
Starting program: /home/user/PrinterToPDF/PrinterConvert 4 3 SIEMENS.C16 1 sdlon ./
[Thread debugging using libthread_db enabled]

[New Thread 0xb7d336c0 (LWP 5434)]

Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0xb7d336c0 (LWP 5434)]
0xb7de99c8 in strcmp () from /lib/libc.so.6
(gdb)

--- 2nd try (end) ---

Ok, I must admit, I was a bit confused where to put the parameters first. But now I know it's with the run command.

Does this help?

**Comment by RWAP** [comment]
Try the latest version - if you get a segmentation fault, use backtrace in gdb to tell you the offending line..

There are problems using the TIFF and Delta Row compression it appears

**Comment by Mike-DE-RE** [comment]
Do you mean like this?

Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0xb7cd16c0 (LWP 8379)]
0xb7d879c8 in strcmp () from /lib/libc.so.6
(gdb) bt
#0  0xb7d879c8 in strcmp () from /lib/libc.so.6
#1  0x0804e057 in main ()

**Comment by Mike-DE-RE** [comment]
When I compile with

gcc PrinterConvert.c -o PrinterConvert -lSDL -lrt -ggdb

and do

gdb PrinterConvert
run 4 3 SIEMENS.C16 1 sdlon ./

I get

Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0xb7e216c0 (LWP 8456)]
0xb7ed79c8 in strcmp () from /lib/libc.so.6
(gdb) bt
#0  0xb7ed79c8 in strcmp () from /lib/libc.so.6
#1  0x0804e057 in main (argc=7, args=0xbf8dab84) at PrinterConvert.c:1808

Better?

**Comment by RWAP** [comment]
You are not using the latest sources from GIthub... If you enter up, that should also show the calls to the error

**Comment by Mike-DE-RE** [comment]
Sorry, do not know why. (Proxy?)

I've reloaded the raw version now and am getting a lot of errors, when doing

gcc PrinterConvert.c -o PrinterConvert -lSDL -lrt -ggdb

I get

PrinterConvert.c:34: warning: data definition has no type or storage class
PrinterConvert.c:34: error: conflicting types for ‘printermemory’
PrinterConvert.c:33: error: previous declaration of ‘printermemory’ was here
PrinterConvert.c:34: warning: initialization makes integer from pointer without a cast
PrinterConvert.c:34: error: initializer element is not constant
PrinterConvert.c: In function ‘initialize’:
PrinterConvert.c:97: warning: comparison between pointer and integer
PrinterConvert.c: In function ‘putpx’:
PrinterConvert.c:307: error: subscripted value is neither array nor pointer
PrinterConvert.c:308: error: subscripted value is neither array nor pointer
PrinterConvert.c:310: error: subscripted value is neither array nor pointer
PrinterConvert.c: In function ‘erasepage’:
PrinterConvert.c:476: error: subscripted value is neither array nor pointer
PrinterConvert.c: In function ‘test_for_new_paper’:
PrinterConvert.c:499: warning: passing argument 4 of ‘write_bmp’ makes pointer from integer without a cast
PrinterConvert.c: In function ‘precedingDot’:
PrinterConvert.c:530: error: subscripted value is neither array nor pointer
PrinterConvert.c: In function ‘_tiff_delta_printing’:
PrinterConvert.c:869: error: subscripted value is neither array nor pointer
PrinterConvert.c: In function ‘main’:
PrinterConvert.c:3549: warning: passing argument 4 of ‘write_bmp’ makes pointer from integer without a cast
PrinterConvert.c:3597: warning: passing argument 1 of ‘free’ makes pointer from integer without a cast

**Comment by RWAP** [comment]
Try that now - I have re-uploaded the source from my PC - it compiled here fine

I think I forgot to hit the SYNC button!

**Comment by Mike-DE-RE** [comment]
Ok, that is much better. Compiles without error and running:

./PrinterConvert 4 3 SIEMENS.C16 1 sdlon ./

delays around ack: t1=0    t2=0    t3=0    t4=0    t5=0
write   = bmp/page1.bmp
command = convert  bmp/page1.bmp  ./pdf/page1.pdf
write   = bmp/page2.bmp
command = convert  bmp/page2.bmp  ./pdf/page2.pdf
write   = bmp/page3.bmp

Converting from bmp to pdf takes ages, no criticism! But that is not the point right now. The result is not as expected.

Here is my test file:

[Test1.zip](https://github.com/RWAP/PrinterToPDF/files/984360/Test1.zip)

Here is the result

[page1.pdf](https://github.com/RWAP/PrinterToPDF/files/984364/page1.pdf)
[page2.pdf](https://github.com/RWAP/PrinterToPDF/files/984365/page2.pdf)

I had to kill the program after page 2 because it took too long. I am quite sure that this can be speeded up a lot. But first of all it has to convert as it should.

Do you want to close this issue and I open a new one?

**Comment by RWAP** [comment]
Open a new issue please - I will mark this one as complete - can you show a screen dump of what the output should actually look like - it looks to be all graphics from a quick look at the file.

**Comment by RWAP** [comment]
The output now seems to be correct  - after some head scratching and de-bugging it was down to casting the integers into floats for the various divisions.
